File: scripts/solar/solar-common/util/lang/HandleUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 9,["7"] = 9,["8"] = 11,["9"] = 11,["10"] = 11,["12"] = 11,["13"] = 17,["14"] = 18,["15"] = 19,["16"] = 19,["17"] = 19,["18"] = 20,["19"] = 19,["20"] = 19,["21"] = 22,["23"] = 24,["24"] = 25,["25"] = 26,["27"] = 28,["28"] = 29,["29"] = 30,["31"] = 32,["32"] = 33,["33"] = 34,["35"] = 36,["36"] = 37,["38"] = 39,["39"] = 17,["40"] = 47,["41"] = 48,["42"] = 49,["43"] = 49,["44"] = 49,["45"] = 50,["46"] = 49,["47"] = 49,["48"] = 52,["50"] = 54,["51"] = 55,["52"] = 56,["53"] = 57,["55"] = 59,["56"] = 60,["57"] = 61,["59"] = 63,["60"] = 64,["61"] = 65,["63"] = 67,["64"] = 47,["65"] = 76,["66"] = 77,["67"] = 78,["68"] = 79,["70"] = 81,["71"] = 76,["72"] = 89,["73"] = 90,["74"] = 91,["75"] = 92,["77"] = 94,["78"] = 89});
local ____exports = {}
local ____SingletonUtil = require("solar.solar-common.util.lang.SingletonUtil")
local SingletonUtil = ____SingletonUtil.default
____exports.default = __TS__Class()
local HandleUtil = ____exports.default
HandleUtil.name = "HandleUtil"
function HandleUtil.prototype.____constructor(self)
end
function HandleUtil.isUnitHandle(self, h)
    if not isDebug then
        SingletonUtil:executeOnce(
            "HandleUtil:isUnitHandle",
            function()
                BJDebugMsg("正式环境无法使用HandleUtil.isUnitHandle")
            end
        )
        return true
    end
    local info = handledef(h)
    if info == nil then
        return false
    end
    if info.reference <= 0 then
        log.errorWithTraceBack("警告：你的Handle引用为0，可能会发生使用过时的handle情况，请排查使用handle的逻辑.")
        return false
    end
    if info.type == nil or info.type == "" then
        log.errorWithTraceBack("警告：你的Handle数据已被底层回收!请实时获取handle,不要使用过时的handle")
        return false
    end
    if info.type == "+w3u" then
        return true
    end
    return false
end
function HandleUtil.isEmptyType(self, h)
    if not isDebug then
        SingletonUtil:executeOnce(
            "HandleUtil:isEmptyType",
            function()
                BJDebugMsg("正式环境无法使用HandleUtil.isEmptyType")
            end
        )
        return true
    end
    local info = handledef(h)
    if info == nil then
        log.errorWithTraceBack("警告：你的Handle定义为空。请检查handle是否正确")
        return true
    end
    if info.reference <= 0 then
        log.errorWithTraceBack("警告：你的Handle引用为0，可能会发生使用过时的handle情况，请排查使用handle的逻辑.")
        return false
    end
    if info.type == nil or info.type == "" then
        log.errorWithTraceBack("警告：你的Handle数据已被底层回收!请实时获取handle,不要使用过时的handle")
        return true
    end
    return false
end
function HandleUtil.getHandleRef(self, h)
    local info = handledef(h)
    if info == nil then
        return 0
    end
    return info.reference
end
function HandleUtil.getHandleType(self, h)
    local info = handledef(h)
    if info == nil then
        return nil
    end
    return info.type
end
return ____exports

File: src/solar/solar-common/util/lang/HandleUtil.ts
/**
 *
 * 此类只能当做debug时排查逻辑 打印信息使用。因为这里返回的信息每个玩家可能不一样 所以不要将此类返回的信息做同步判断执行需要同步操作的逻辑
 *
 * handle 数量和大部分信息 是不能异步的。但handledef返回的信息可能是异步的(受限于不同玩家的handle使用和底层销毁时间??)
 * 注意:可能每个玩家返回的数据不一样。请不要当做同步数据使用 以免异步
 * @param h
 */
import SingletonUtil from "@/SingletonUtil";

export default class HandleUtil {

    /**
     * 检查是否是单位的handle
     * @param h
     */
    static isUnitHandle(h: unit | handle): boolean {
        if (!isDebug) {
            SingletonUtil.executeOnce("HandleUtil:isUnitHandle", () => {
                BJDebugMsg("正式环境无法使用HandleUtil.isUnitHandle")
            })
            return true;
        }
        let info = handledef(h);
        if (info == null) {
            return false;
        }
        if (info.reference <= 0) {
            log.errorWithTraceBack("警告：你的Handle引用为0，可能会发生使用过时的handle情况，请排查使用handle的逻辑.")
            return false;
        }
        if (info.type == null || info.type == "") {
            log.errorWithTraceBack("警告：你的Handle数据已被底层回收!请实时获取handle,不要使用过时的handle")
            return false;
        }
        if (info.type == "+w3u") {
            return true;
        }
        return false;
    }

    /**
     * 检查是否是空类型的handle(即底层已回收此handle对应的数据)
     * 只能用于debug时使用
     * @param h
     */
    static isEmptyType(h: handle): boolean {
        if (!isDebug) {
            SingletonUtil.executeOnce("HandleUtil:isEmptyType", () => {
                BJDebugMsg("正式环境无法使用HandleUtil.isEmptyType")
            })
            return true;
        }
        let info = handledef(h);
        if (info == null) {
            log.errorWithTraceBack("警告：你的Handle定义为空。请检查handle是否正确")
            return true;
        }
        if (info.reference <= 0) {
            log.errorWithTraceBack("警告：你的Handle引用为0，可能会发生使用过时的handle情况，请排查使用handle的逻辑.")
            return false;
        }
        if (info.type == null || info.type == "") {
            log.errorWithTraceBack("警告：你的Handle数据已被底层回收!请实时获取handle,不要使用过时的handle")
            return true;
        }
        return false;
    }


    /**
     * 获取handle的 引用计数
     * （正常使用的handle 应该大于0 否则此handle即为已销毁的handle 会被底层重新分配）
     * @param h
     */
    static getHandleRef(h: handle): number {
        let info = handledef(h);
        if (info == null) {
            return 0;
        }
        return info.reference;
    }

    /**
     * 获取handle的类型
     * (返回 +w3u等等)
     * @param h
     */
    static getHandleType(h: handle): string {
        let info = handledef(h);
        if (info == null) {
            return null;
        }
        return info.type;
    }
}